### ISSUE TITLE
`unused_assignments` does not trigger on assignment to reference to temporary (lifetime extension)

### ISSUE THREAD

**Issue by briansmith** [rust-lang/rust#144031]
### Code

```Rust
fn store32(output: &mut [u8; 16], i: usize, value: u32) {
    let output: &mut [u8; 4] = &mut output[i..][..4].try_into().unwrap();
    *output = u32::to_le_bytes(value);
}

fn main() {
    let mut r = [0u8; 16];
    store32(&mut r, 0, 0x12345678);
    println!("{r:?}")
}
```

### Current output

```Shell
None
```

### Desired output

```Shell
warning: value assigned to `output` is never read
note: `#[warn(unused_assignments)]` on by default


The fix is:

-    let output: &mut [u8; 4] = &mut output[i..][..4].try_into().unwrap();
+    let output: &mut [u8; 4] = (&mut output[i..][..4]).try_into().unwrap();


The compiler should have recognized that `output` in the buggy version is a reference to a temporary. The temporary is written (through `output`) but never read.
```

### Rationale and extra context

The author of this code overlooked that `.try_into().unwrap()` binds tighter than `&mut `, so the `store32` function is buggy: It doesn't write anything to `output` like it intends. It doesn't make sense to modify a temporary and then never read from it.

### Other cases

```Rust

```

### Rust Version

```Shell
% rustc --version --verbose
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: aarch64-apple-darwin
release: 1.88.0
LLVM version: 20.1.5
```

### Anything else?

See the thread https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Accidental.20borrow.20of.20temporary.2Faccidental.20lifetime.20extension/with/528940661

**Comment by hanna-kruppe** [comment]
Lifetime extension and temporaries may be red herrings as far as the lint implementation is concerned, I don't think the relevant lints currently try to reason about *any* assignment through references. For example, this doesn't trigger any lint either:

```rust
pub fn foo() {
    let mut x = 0;
    let r: &mut i32 = &mut x;
    *r = 1;
}
```

**Comment by briansmith** [comment]
See also #44340, though #44340 is slightly different because the test case is a store through a reference that is immediately overwritten by another store through that reference, such that the first store could be optimized away, but where the stores could have observable effects. Whereas this issue is about the case where there are no observable effects.

**Comment by briansmith** [comment]
See also the example by @nazar-pc  at https://github.com/rust-lang/rust/issues/144645#issue-3274439956, where `MaybeUninit::write` is used to the same ill effect:

```rust
#![feature(maybe_uninit_uninit_array_transpose)]

#[derive(Debug)]
struct S {
    a: [[u8; 32]; 2],
}

use std::mem::MaybeUninit;
fn main() {
    let mut s = Box::<S>::new_uninit();
    let s_ptr = s.as_mut_ptr();
    let a_ptr = unsafe { &raw mut (*s_ptr).a };
    let a = unsafe {
        &mut *a_ptr.cast::<[MaybeUninit<[u8; 32]>; 2]>()
    };
    a.transpose().write([[1u8; 32]; 2]);     // <-- write to a temporary.
    dbg!(unsafe { s.assume_init() });
}
```
